hunts/src/HuntViewMain.tsx

```diff
@@ -41,32 +33,79 @@ const MODAL_COMPONENTS: Record<string, any> = {
   EDIT_TAGS: EditPuzzleTagsModal,
 };
 
-export const HuntViewMain = (props: { huntId: HuntId }) => {
-  const tableData = useSelector<RootState, any>(selectPuzzleTableData);
-  const modal = useSelector<RootState, any>((state) => state.modal);
-  const hunt = useSelector<RootState, any>((state) => state.hunt);
-  const alert = useSelector<RootState, any>((state) => state.alert);
-  const filterTags = useSelector<RootState, PuzzleTag[]>(getTagFilter);
+export interface PuzzleTable extends Puzzle {
+  subRows: Puzzle[];
+}
 
-  const filterSolved = useSelector(getSolveStateFilter);
+const makePuzzleTableData = (ids: Puzzle[]) => {
+  // We need to construct the meta/subtree relationships.
+  // Each meta that contains a puzzle gets a reference to the same
+  // puzzle object. Then we can offload any actual graph traversal
+  // to the table library.
 
-  const dispatch = useDispatch<Dispatch>();
+  // Make a deep copy of everything first
+  const rowsCopy: PuzzleTable[] = ids.map((id) => ({
+    ...id,
+    subRows: [],
+  }));
+  const rowMap: { [id: PuzzleId]: Puzzle } = {};
+  rowsCopy.forEach((row) => {
+    rowMap[row.id] = row;
+  });
+  // First give every meta references to all its children
+  rowsCopy.forEach((row) => {
+    if (row.feeders.length > 0) {
+      row.subRows = [];
+      row.feeders.forEach((subRowId) => {
+        // This check is needed to deal with inconsistent data:
+        // if we just deleted a puzzle, it may still appear as a feeder for
+        // another puzzle if we haven't done a full refresh of the data yet
+        if (subRowId in rowMap) {
+          row.subRows.push(rowMap[subRowId]);
+        }
+      });
+    }
+  });
+  // Once all the meta-child relationships are there, we can sort
+  // every list of puzzles: all the subRows lists, and the list of outer puzzles as well.
+  rowsCopy.forEach((row) => {
+    if (row.subRows) {
+      row.subRows.sort(puzzleComparator);
+    }
+  });
+  const outerRows = rowsCopy.filter((row) => row.metas.length == 0);
+  outerRows.sort(puzzleComparator);
+  return outerRows;
+};
+
+export const HuntViewMain = (props: { huntId: HuntId }) => {
+  const modal = useStore((state) => state.modalSlice);
+  const hunt = useStore((state) => state.huntSlice);
+  const alert = useStore((state) => state.alertSlice);
+  const { fetchAllPuzzles, puzzles } = useStore((state) => state.puzzlesSlice);
+  const { tags: filterTags, toggleFilterTag } = useStore(
+    (state) => state.filterSlice
+  );
+  const filterSolved = useStore((state) => state.filterSlice.solveStateFilter);
+  const { hideAlert } = useStore((state) => state.alertSlice);
 
-  const updatePuzzleData = () => {
-    dispatch(fetchPuzzles(props.huntId));
-  };
+  const l = useStore((state) => state.puzzlesSlice.lastUpdate);
 
-  useInterval(updatePuzzleData, 10 * 1000);
+  useInterval(() => {
+    fetchAllPuzzles();
+  }, 10 * 1000);
```

Comment:
Oh, maybe one more of these calls that can be collapsed into `useInterval(fetchAllPuzzles, 10 * 1000);`?